File: java/src/main/java/com/github/propi/rdfrules/java/Transformable.java
package com.github.propi.rdfrules.java;

import java.util.function.Predicate;
import java.util.function.UnaryOperator;

/**
 * Created by Vaclav Zeman on 2. 5. 2018.
 */
public interface Transformable<T, SColl, JColl> {

    com.github.propi.rdfrules.data.ops.Transformable<T, SColl> asScala();

    JColl asJava(SColl scala);

    default JColl map(UnaryOperator<T> f) {
        return asJava(asScala().map(f::apply));
    }

    default JColl filter(Predicate<T> f) {
        return asJava(asScala().filter(f::test));
    }

    default JColl slice(int from, int until) {
        return asJava(asScala().slice(from, until));
    }

    default JColl take(int n) {
        return asJava(asScala().take(n));
    }

    default JColl drop(int n) {
        return asJava(asScala().drop(n));
    }

    default int size() {
        return asScala().size();
    }

}

File: core/src/main/scala/com/github/propi/rdfrules/data/ops/Transformable.scala
package com.github.propi.rdfrules.data.ops

import com.github.propi.rdfrules.utils.ForEach

/**
  * Created by Vaclav Zeman on 2. 2. 2018.
  */
trait Transformable[T, Coll] {

  protected def coll: ForEach[T]

  protected def transform(col: ForEach[T]): Coll

  def map(f: T => T): Coll = transform(coll.map(f))

  def flatMap(f: T => ForEach[T]): Coll = transform(coll.flatMap(f))

  def filter(f: T => Boolean): Coll = transform(coll.filter(f))

  def filterIndices(indices: Set[Int]): Coll = transform(coll.zipWithIndex.filter(x => indices(x._2)).map(_._1))

  def slice(from: Int, until: Int): Coll = transform(coll.slice(from, until))

  def take(n: Int): Coll = transform(coll.take(n))

  def drop(n: Int): Coll = transform(coll.drop(n))

  def head: T = coll.head

  def headOption: Option[T] = coll.headOption

  def find(f: T => Boolean): Option[T] = coll.find(f)

  lazy val size: Int = coll.size

}